src/Purebred/Mail.hs

```diff
@@ -0,0 +1,55 @@
+-- This file is part of purebred
+-- Copyright (C) 2019 Róman Joost
+--
+-- purebred is free software: you can redistribute it and/or modify
+-- it under the terms of the GNU Affero General Public License as published by
+-- the Free Software Foundation, either version 3 of the License, or
+-- (at your option) any later version.
+--
+-- This program is distributed in the hope that it will be useful,
+-- but WITHOUT ANY WARRANTY; without even the implied warranty of
+-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+-- GNU Affero General Public License for more details.
+--
+-- You should have received a copy of the GNU Affero General Public License
+-- along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+{- |
+= Intention
+
+This module provides functionality to access data and convert
+Data.MIME instances using a unified interface.
+-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE FlexibleInstances #-}
+module Purebred.Mail
+  ( MIMEData(..)
+  ) where
+
+import Data.Maybe (fromMaybe)
+import Control.Lens (filtered, has, preview, view)
+import qualified Data.Text as T
+
+import UI.Utils (takeFileName)
+import Data.MIME
+
+-- | Provide metadata used for render the message using a unified
+-- interface
+class MIMEData a where
+  contentType :: a -> T.Text
+  filename :: CharsetLookup -> a -> T.Text
+  attachmentType :: a -> T.Text
```

Comment:
I'm really confused as to why this abstraction is necessary.  All of these only read from the headers, right?
`MIMEMessage` and `WireEntity` are both type synonyms for `Message something something'`, so you should be able to use *exactly* the same functions to read these data from either type.

Reply:
hrm.. odd.. I'll have another look. Cheers!